MovieQuiz/Presentation/MovieQuizPresenter.swift

```diff
@@ -0,0 +1,142 @@
+import UIKit
+
+final class MovieQuizPresenter: QuestionFactoryDelegate {
+    
+    // MARK: - Properties
+    
+    private let statisticService: StatisticService!
+    private var questionFactory: QuestionFactoryProtocol?
+    weak var viewController: MovieQuizViewControllerProtocol?
+
+    private var currentQuestion: QuizQuestion?
+    private let questionsAmount: Int = 10
+    private var currentQuestionIndex: Int = 0
+    private var correctAnswers: Int = 0
+    
+    init(viewController: MovieQuizViewControllerProtocol) {
+        self.viewController = viewController
+        self.statisticService = StatisticServiceImplementation()
+        self.questionFactory = QuestionFactory(moviesLoader: MoviesLoader(), delegate: self)
+        self.questionFactory?.loadData()
+        viewController.showLoadingIndicator()
+        }
+    
+    //MARK: - Func
+    
+    func didLoadDateFromServer() {
+            viewController?.hideLoadingIndicator()
+            questionFactory?.requestNextQuestion()
+        }
+        
+    func didFailToLoadData(with error: Error) {
+            let massage = error.localizedDescription
+            viewController?.showNetworkError(message: massage)
+        }
+    
+    func didRecieveNextQuestion(question: QuizQuestion?) {
+            guard let question = question else {
+                return
+            }
+
+            currentQuestion = question
+            let viewModel = convert(model: question)
+            DispatchQueue.main.async { [weak self] in
+                self?.viewController?.show(quiz: viewModel)
+            }
+        }
+    
+    func didAnswer(isCorrectAnswer: Bool) {
+            if isCorrectAnswer {
+                correctAnswers += 1
+            }
+        }
+    
+    func isLastQuestion() -> Bool {
+        currentQuestionIndex == questionsAmount - 1
+    }
+    
+    func resetQuestionIndex() {
+        currentQuestionIndex = 0
+    }
+    
+    func switchToNextQuestion() {
+        currentQuestionIndex += 1
+    }
+    
+    func convert(model: QuizQuestion) -> QuizStepViewModel {
+        QuizStepViewModel(
+            image: UIImage(data: model.image) ?? UIImage(),
+            question: model.text,
+            questionNumber: "\(currentQuestionIndex + 1)/\(questionsAmount)"
+        )
+    }
+    
+    private func proceedToNextQuestionOrResults() {
+            if self.isLastQuestion() {
```

Comment:
self можно опускать вне замыканий
```suggestion
            if isLastQuestion() {
```